src/base/utils/python.cpp

```diff
@@ -0,0 +1,128 @@
+ * exception statement from your version.
+ */
+
+#include "python.h"
+
+#include <QCoreApplication>
+#include <QProcess>
+#include <QStringList>
+
+#include "base/logger.h"
+#include "base/utils/version.h"
+
+struct PythonInfo
+{
+    using Version = Utils::Version<qint8, 3, 1>;
+
+    bool isEmpty() const
+    {
+        return (executableName.isEmpty() || (version == Version {}));
+    }
+
+    QString executableName;
+    Version version {};
+};
+
+namespace
+{
+    PythonInfo pythonInfo()
+    {
+        static PythonInfo pyInfo;
+        if (pyInfo.isEmpty()) {
+            const auto testPythonInstallation = [](const QString &exeName) -> bool
+            {
+                QProcess proc;
+                proc.start(exeName, {"--version"}, QIODevice::ReadOnly);
+                if (proc.waitForFinished() && (proc.exitCode() == QProcess::NormalExit)) {
+                    QByteArray procOutput = proc.readAllStandardOutput();
+                    if (procOutput.isEmpty())
+                        procOutput = proc.readAllStandardError();
+                    procOutput = procOutput.simplified();
+
+                    // Software 'Anaconda' installs its own python interpreter
+                    // and `python --version` returns a string like this:
+                    // "Python 3.4.3 :: Anaconda 2.3.0 (64-bit)"
+                    const QList<QByteArray> outputSplit = procOutput.split(' ');
+                    if (outputSplit.size() <= 1)
+                        return false;
+
+                    pyInfo.executableName = exeName;
+                    pyInfo.version = outputSplit.at(1);
+                    LogMsg(QCoreApplication::translate("python", "Python detected, version: %1").arg(pyInfo.version), Log::INFO);
+                    return true;
+                }
+
+                return false;
+            };
+
+#if defined(Q_OS_UNIX)
+            // On Unix-Like Systems python2 and python3 should always exist
+            // https://legacy.python.org/dev/peps/pep-0394/
+            if (testPythonInstallation("python3"))
+                return pyInfo;
+            if (testPythonInstallation("python2"))
+                return pyInfo;
+#endif
+            // Look for "python" in Windows and in UNIX if "python2" and "python3" are
+            // not detected.
+            if (testPythonInstallation("python"))
+                return pyInfo;
+
+            LogMsg(QCoreApplication::translate("python", "Python not detected"), Log::INFO);
+        }
+
+        return pyInfo;
+    }
+}
+
+/**
+ * Returns the major version number from `python --version`.
+ */
+int Utils::Python::majorVersion()
+{
+    return pythonInfo().version.majorNumber();
+}
+
+/**
+ * Returns the python executable name, e.g. "python3".
+ */
+QString Utils::Python::executableName()
+{
+    return pythonInfo().executableName;
+}
+
+/**
+ * Returns python version in 3-component form, e.g. "2.7.9".
+ */
+QString Utils::Python::fullVersion()
```

Comment:
I would implement this in the Version class.

Reply:
I plan to submit another PR for this and other refactors, so not now.

Reply:
Are you sure about that? The function is just a stripped down version of the `operator QString()`:
```c++
QString fullVersionString() const
{
	QString res = QString::number((*this)[0]);
	for (std::size_t i = 1; i <= N; ++i)
		res += QLatin1Char('.') + QString::number((*this)[i]);
	return res;
}
```

Reply:
>Are you sure about that? 

Oops I intend to reply to the [comment](https://github.com/qbittorrent/qBittorrent/pull/9113#discussion_r196489192) above.

>I would implement this in the Version class.

OK, I'll update PR later.

Reply:
Or even simpler:
```c++
QString fullVersionString() const
{
    return static_cast<QString>(Version<T,N,N>(m_components));
}
```

Reply:
>Or even simpler:

The class doesn't expose its underlying type, so can't be done, this will work instead:
`return Version<T, N, N> {static_cast<QString>(*this)};`
But this is not very efficient.

Reply:
Ah, indeed... Given the other constructor that takes `Components...`, I think the following looks reasonable:
```diff
diff --git a/src/base/utils/version.h b/src/base/utils/version.h
index 19213cb35..fe6832351 100644
--- a/src/base/utils/version.h
+++ b/src/base/utils/version.h
@@ -62,6 +62,12 @@ namespace Utils
         {
         }
 
+        template <typename Other>
+        Version(const std::array<Other, N>& versionParts)
+            : m_components {versionParts}
+        {
+        }
+
         /**
          * @brief Creates version from string in format "x.y.z"
          *
@@ -129,6 +135,11 @@ namespace Utils
             return res;
         }
 
+        QString fullVersionString() const
+        {
+            return static_cast<QString>(Version<T,N,N>(m_components));
+        }
+
         constexpr bool operator==(const ThisType &other) const
         {
             return (m_components == other.m_components);
```

Reply:
Oh, the `static_cast` is unneeded (mistakenly assumed that the operator QString is explicit).

Reply:
Regarding the failing build with GCC 4.8, I think you might help it replacing `{}` with `()` (see https://github.com/Chocobo1/qBittorrent/blob/d61c1e0cd4147e15cfc6c260f38b31d700674e9a/src/base/utils/version.h#L192).